lib/Echidna/Output/JSON.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Echidna.Output.JSON where
+
+import Echidna.ABI (ppAbiValue)
+import qualified Echidna.Campaign as C
+import Echidna.Solidity (SolTest)
+import Echidna.Transaction (Tx(..), TxCall(..))
+import Data.Aeson hiding (Error)
+import qualified Data.ByteString.Base16 as BS16
+import Data.ByteString.Lazy (ByteString)
+import Data.Text
+import Data.Text.Encoding (decodeUtf8)
+
+data Top = Top
+  { _success :: Bool
+  , _error :: Maybe String
+  , _tests :: [Test]
+  , seed :: Int
+  }
+
+instance ToJSON Top where
+  toJSON Top{..} = object
+    [ "success" .= _success
+    , "error" .= _error
+    , "tests" .= _tests
+    , "seed" .= seed
+    ]
+
+data Test = Test
+  { contract :: Text
+  , name :: Text
+  , status :: TestStatus
+  , _error :: Maybe String
+  , testType :: TestType
+  , transactions :: Maybe [Transaction]
+  }
+
+instance ToJSON Test where
+  toJSON Test{..} = object
+    [ "contract" .= contract
+    , "name" .= name
+    , "status" .= status
+    , "error" .= _error
+    , "type" .= testType
+    , "transactions" .= transactions
+    ]
+
+data TestType = Property | Assertion
+
+instance ToJSON TestType where
+  toJSON Property = "property"
+  toJSON Assertion = "assertion"
+
+data TestStatus = Fuzzing | Shrinking | Solved | Passed | Error
+
+instance ToJSON TestStatus where
+  toJSON Fuzzing = "fuzzing"
+  toJSON Shrinking = "shrinking"
+  toJSON Solved = "solved"
+  toJSON Passed = "passed"
+  toJSON Error = "error"
+
+
+data Transaction = Transaction
+  { contract :: Text
+  , function :: Text
+  , arguments :: Maybe [String]
+  , gas :: Integer
+  , gasprice :: Integer
+  }
+
+instance ToJSON Transaction where
+  toJSON Transaction{..} = object
+    [ "contract" .= contract
+    , "function" .= function
+    , "arguments" .= arguments
+    , "gas" .= gas
+    , "gasprice" .= gasprice
+    ]
+
+encodeCampaign :: C.Campaign -> Int -> ByteString
+encodeCampaign C.Campaign{_tests} seed = encode
+  -- TODO: success and error are hardcoded, add after reworking error reporting
+  Top { _success = True
+      , _error = Nothing
+      , _tests = mapTest <$> _tests
+      , seed = seed
+      }
+
+mapTest :: (SolTest, C.TestState) -> Test
+mapTest (solTest, testState) =
+  let (status, transactions, err) = mapTestState testState in
+  Test { contract = "" -- TODO add when mapping is available
```

Comment:
Convert this TODO into a github issue. If you are going to work on that, don't forget to assign it to yourself. :smile:

Reply:
done